Simplify access functions by introducing chainable access rules

Merseleo opened the pull request:
Closes #144 

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a service, which does x -->

- I added a generic interface `AccessRule` and the two generic functions `orElse` and `andAlso` that allow to concatenate certain access checks, such as whether a user is an admin, and chain them together to build more complex rules
  - moreover, you can use `orElseThrow` to specify a certain exception that should be thrown if the chain until the point of this operation does not succeeded
  - the interface is designed to only build rule chains, they must be evaluated using `checkFor` which calls the rules with the provided requesting user object and the provided target entity (or Unit, if no target is accessed, but only some roles are checked)
  - I added some further helper functions to chain together rules that does not operate directly on the same entity type but instead on a more generic (e.g. isServerAdmin can be called on every entity type)
- I added a concrete implementation `UserAccessRule` that contain frequently used access check functions, such as whether a user is an admin or whether two users are actually the same user and integrated these rules into the user service
- Similarly, I added access rules for reviews, project and criteria and refactored all services.

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have adapted the tests
- [x] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 commented:
In GitLab by @dts84 on Apr 27, 2022, 22:01

mentioned in commit 16963e08940ec8aa43d198bce11ad0cd5a6a0635

Commit:
feat: add a generic access rule interface to define access checks that could be chained together

Additionally, add some basic rules for checking the access of a user to demonstrate the usage.

Commit:
refactor: integrate new (user) access rules into the user service.

The 'updateUser' did not get the update yet.

Adds further user access rules and updates the structure of the general AccessRule. An exception must not be the terminal operation but instead chained rules (including exceptions) could be build and then evaluated using the 'checkFor' function.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/AccessRule.kt:
```diff
@@ -0,0 +1,62 @@
+ * operations, such as: fetching data from a database (e.g., project memberships).
+ * By making every access rule `suspend`, we can easily chain rules that actually need to
+ * suspend with rules that does not necessarily need to suspend without any checking, whether
+ * one rule in the chain needs to suspend. Moreover, it no data need to be fetcher or similar tasks
```
```suggestion
 * one rule in the chain needs to suspend. Moreover, if no data need to be fetcher or similar tasks
```

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/UserAccessRule.kt:
```diff
@@ -0,0 +1,82 @@
+/**
+ * Verifies that the [user] has the role [UserRole.USER_ROLE_ADMIN].
+ *
+ * If the user is not a server admin, a [se.uulm.snowballr.backend.model.SnowballRException.UnauthorizedException] is thrown.
```
Replace the `se.uulm.snowballr.backend.model.SnowballRException.UnauthorizedException` with `UnauthorizedException` and import it.

Merseleo replied:
You are right, I replaced the imports.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/UserAccessRule.kt:
```diff
@@ -0,0 +1,82 @@
+ * If the user is not a server admin, a [se.uulm.snowballr.backend.model.SnowballRException.UnauthorizedException] is thrown.
+ *
+ * @param user The user to verify
+ * @param getException Getter method for the subtype of [se.uulm.snowballr.backend.model.SnowballRException.UnauthorizedException], which is thrown
```
Same here for the `UnauthorizedException`.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -133,78 +143,82 @@ class UserService(
-        ) {
-            throw NotFoundException(EntityType.USER, request.id)
-        }
+        // TODO: question for reviewer: What do you think, may a server admin request a deleted user by its id?
```
I am not sure if it really is useful, but imo we can allow this, because this should only be of statistical relevance. If a fully deleted user is returned, the only information the server admin gets is, that once  a user existed in this line. Thus, I don't see a reason why we should deny the access for that. What do you think?

Merseleo replied:
Ok, i have the same opinion, I only wanted to know whether I missed a possible problem.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/AccessRule.kt:
```diff
@@ -47,16 +49,49 @@ fun <T> AccessRule<T>.andAlso(next: AccessRule<T>): AccessRule<T> = AccessRule {
-suspend fun <T> AccessRule<T>.orElseThrow(requester: User, target: T, exception: SnowballRException) {
-    if (!this.isAllowedToAccess(requester, target)) {
-        throw exception
+suspend fun <T> AccessRule<T>.checkFor(requester: User, target: T) {
+    val isAllowedToAccessEntity = this.isAllowedToAccess(requester, target)
+
+    if (!isAllowedToAccessEntity) {
+        throw AccessRuleCheckFailedException()
```
As far as I understand you implemented this method, to check, whether an exception is correctly thrown or not and if not the `AccessRuleCheckFailedException` is thrown. Does this even makes sense then? Because now we have to add the `.checkFor` after every rule chain, otherwise the `AccessRuleCheckFailedException` is not thrown either. So the whole check for exceptions only works, if we don't forget this, which makes the check kind of useless imo, or do I get something wrong?

I get what you are trying to do here. We should have a obligatory throw-rule at the end of every rule chain in case it fails, but at the moment I don't know how to implement this.

Merseleo replied:
No, this wasn't the intention of the `checkFor` function. Actually I added it to separate the check + exception from the entities to be checked. Before I introduced this function, you had to provide the requesting user and the requested entity for every "terminating" function as `orElseThrow`. This way we separate it and make the code more readable imo. 
And yes, we have to add the `checkFor` statement at the end of every chain, but we have to provide the requester and target anyway, either in a `orElseThrow` or in a `checkFor`. The last check with the `AccessRuleCheckFailedException` is only an add-on. 

But I am not sure, what is the best solution here. But your suggestion with the obligatory throw rule is the `checkFor` function, isn't it? 
Maybe we should discuss this in person.

domienderle replied:
As we discussed, please change the rules to use the annotation, so we don't forget to call the `checkFor` when using the rules.

Merseleo replied:
I will consider this whenever I write the new rules now.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/AccessRule.kt:
```diff
@@ -47,16 +49,49 @@ fun <T> AccessRule<T>.andAlso(next: AccessRule<T>): AccessRule<T> = AccessRule {
+    this.isAllowedToAccess(requester, Unit)
+}
+
+fun <TNew, TOld> AccessRule<TOld>.forProperty(prop: KProperty1<TNew, TOld>): AccessRule<TNew> =
+    AccessRule { requester, newTarget -> this.isAllowedToAccess(requester, prop.get(newTarget)) }
```
Write a comment for this function.

Merseleo replied:
Sorry, forgot this method. Otherwise it is quite difficult to understand, I know.

Merseleo replied:
What do you think of the new comments? Is it more understandable now?

domienderle replied:
Yes this comment is really good, thank you. Only one minor thing. Add a link qualifier to `[java.util.UUID]` to the current `[UUID]` or import `java.util`, so it links correctly.

Merseleo replied:
Thank you for the hint, didn't notice that.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/UserAccessRule.kt:
```diff
@@ -48,7 +63,9 @@ fun isInSameProject(projectMemberRepo: IProjectMemberTableRepo): UserAccessRule
-    isServerAdmin.orElseThrow(user, user, getException(user.id.toString()))
+    isServerAdmin.forTarget<User>()
+        .orElseThrow(getException(user.id.toString()))
+        .checkFor(user, user)
```
Isn't this check useless? Overall, aren't this checks useless at all if we have a `orElseThrow` implemented?

Merseleo replied:
No, the `checkFor` function injects the user into the rule chain? Otherwise we would have build a chain without running it. Maybe you should consider it as the builder pattern. `isServerAdmin.forTarget<User>().orElseThrow(getException(user.id.toString()))` builds a chain and returns a `AccessRule<User>` (the builder). With `checkFor` you evaluate this chain for a certain input to get either the Unit type (so you know, that your assertions holds) or an exception is thrown (and you know, that this access was not allowed).

domienderle replied:
With our discussed results this conversation is not up-to-date anymore and can be resolved.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/AccessRule.kt:
```diff
@@ -47,16 +49,49 @@ fun <T> AccessRule<T>.andAlso(next: AccessRule<T>): AccessRule<T> = AccessRule {
+ *
+ * @return An [AccessRule] for the target type [T] that applies the same logic.
+ */
+fun <T> AccessRule<Unit>.forTarget(): AccessRule<T> = AccessRule { requester, _ ->
```
Is `forTarget` here the correct name? I am not sure if I really get the functionality of this method correct, but wouldn't be something like `ignoreTarget` be better?

Merseleo replied:
You are right, actually the functionality is `ignoreTarget`, but imo with `forTarget` you get a better readability: `isServerAdmin.forTarget<User>()` vs `isServerAdmin.ignoreTarget<User>()`. In general, I (with the massive help of AI) decided to use names that make the code more readable. And if you want to understand exactly, what the function does, then you have an explaining comment (I improved it).

domienderle replied:
I really like the comment of the method, but I'm still no fan of the name, because imo the code is still well readable with `ignoreTarget`. `forTarget` imo indicates the complete opposite of the methods functionality. But we can also ask the others if you want.

Merseleo replied:
Yes, I would ask the others as well and if they decide too to use `ignoreTarget`, it is a simple automatic IDE renaming.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -133,78 +143,82 @@ class UserService(
         val targetUserId = parseUUID(request.id, EntityType.USER)
 
-        verifyUserAccess(currentUser, targetUserId, IdentifierType.ID)
+        isAllowedToReadUser(currentUser, targetUserId, IdentifierType.ID).checkFor(currentUser, targetUserId)
```
Isn't `checkFor` useless here, since it is always true or if an exception in `isAllowedToReadUser` is thrown it is not executed anymore.

Merseleo replied:
Check my previous comments.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/AccessRule.kt:
```diff
@@ -0,0 +1,97 @@
+ * This interface is `suspend` as some access checks require asynchronous
+ * operations, such as: fetching data from a database (e.g., project memberships).
+ * By making every access rule `suspend`, we can easily chain rules that actually need to
+ * suspend with rules that does not necessarily need to suspend without any checking, whether
```
```suggestion
 * suspend with rules that do not necessarily need to suspend without any checking, whether
```

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/AccessRule.kt:
```diff
@@ -0,0 +1,97 @@
+
+/**
+ * Checks whether this rule holds; if so, the next rule is checked.
+ * Otherwise, the given exception is thrown and the chains terminates.
```
```suggestion
 * Otherwise, the given exception is thrown and the chain terminates.
```

domienderle reviewed (changes requested):
I really like the approach of implementing rules that can be chained together. Your implementation is really good - well done - but not easy to understand at the first glance. 

One major thing I don't understand is why we need `checkFor` if we have `orElseThrow`. Imo the problem at the moment is, that we could end chains without having an exception. The apporach to avoid this by having the `checkFor` function only shifts this problem, that we have to call the `checkFor` function. I am not sure how this could be implemented, but we should have an obligatory `orElseThrow` at the end of every rule, instead of manually calling check functions.

Commit:
docs: add, correct or improve comments for the access rule system

Merseleo reviewed (commented):
Ok, thank you for the review. I fixed all problems you found. 
As we discussed, if you do not have the time to finish this review, maybe somebody else can finish it after the refactoring on the other services.

domienderle reviewed (changes requested):
I had a look on our previous conversations. So far most of them look good and could be resolved. Only a couple need some rework.

Commit:
feat: overload 'orElseThrow' function to accept an exception provider and not an exception directly.

This can be used if the exception needs either the requesting user or the target entity to be constructed.

Merseleo reviewed (commented):
Thank you for the quick feedback again. I will now rebase and finish the remaining access checks and then request a review again.

Commit:
refactor: extract fetching of authors and backward references and converting the paper to a grpc paper into a function 'toGrpcPaperWithAuthorsAndBackwardReferences'.

Additionally, this function is directly used in the paper service.

Commit:
refactor: use 'toGrpcPaperWithAuthorsAndBackwardReferences' in the reading list service and improve comments

Commit:
refactor: enhance usage of user access rules and integrate the rules into the 'updateUser' call

Moreover, some additional checks were integrated that were missing (+ the tests adapted) and some more specific exceptions were added if they were used multiple times.

Commit:
refactor: add project access rules and integrate them into the project service.

Additionally, remove unnecessary authorization helper.

Commit:
refactor: add review access rules and integrate them into the review service.

Commit:
refactor: integrate project access rules into the invitation service.

Additionally, improve exception clarity.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -98,77 +108,51 @@ class UserService(
+        isServerAdmin.forTarget<User>()
+            .orElse(isTargetUserActive)
+            .orElseThrow(UserNotFoundException(request.id))
+            .checkFor(currentUser, targetUser)
```
Why do we check here for `isServerAdmin` again? We do this already in `isAllowedToReadUser`.

Shouldn't this be
```kotlin
        isTargetUserActive.forTarget<User>()
            .orElseThrow(UserNotFoundException(request.id))
            .checkFor(currentUser, targetUser)
```

Merseleo replied:
No. Actually it could be that the user is not a server admin but only in the same project. Then the `isAllowedToReadUser` check passes, but the second one do not pass, if the requested user is deleted, because then it is not enough to be in the same project, but you have to be a server admin.

Merseleo replied:
So your suggestion would also work, but allow access to the user in scenarios we actually do not want (or at least defined as scenarios that are not allowed)

domienderle replied:
Yes I agree, we have to check this. First I had the same thought as @Slartibartfass2, but @Merseleo explanation is correct.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -98,77 +108,51 @@ class UserService(
+            .andAlso(
+                isServerAdmin.forTarget<User>()
+                    .orElse(isTargetUserActive)
+                    .orElseThrow(UserNotFoundException(request.email, IdentifierType.EMAIL)),
```
Same here. We already check for the user being a server admin, right?

Merseleo replied:
Same answer.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -98,77 +108,51 @@ class UserService(
-        verifyServerAdminRole(currentUser) { UnauthorizedException.All(EntityType.USER, AccessType.READ, it) }
+        isServerAdmin.forTarget<User>()
+            .orElseThrow(UnauthorizedException.All(EntityType.USER, AccessType.READ, currentUser.id.toString()))
+            .checkFor(currentUser, currentUser)
```
Could we add a short form when we don't use the target?

This could be
```suggestion
        isServerAdmin
            .orElseThrow(UnauthorizedException.All(EntityType.USER, AccessType.READ, currentUser.id.toString()))
            .checkFor(currentUser)
```

Merseleo replied:
I already simplified it a bit, but I agree, maybe I can introduce a single parameter `checkFor`.

Slartibartfass2 replied:
I think you can add a default value to the second parameter.

domienderle replied:
You cannot add a default value for the second parameter since it is generic. What do you think of overloading the `checkFor` function like this?
```kotlin
suspend fun AccessRule<Unit>.checkFor(requester: User) {
    val allowed = isAllowedToAccess(requester, Unit)
    if (!allowed) throw AccessRuleCheckFailedException()
}
```

Slartibartfass2 replied:
Good catch @domienderle, I didn't check whether this would work.

We can simply add
```kotlin
suspend fun AccessRule<Unit>.checkFor(requester: User) = checkFor(requester, Unit)
```

Merseleo replied:
Ok, then I will add this overloaded function.

Commit:
refactor: integrate project paper access rules into the project paper service.

Additionally, improve helper functions in the project paper service and remove duplicate code.

Merseleo reviewed (commented):
@Slartibartfass2 Thank you for your ideas.

Commit:
refactor: add criterion access rules and integrate them into the criterion service.

Merseleo reviewed (commented):
Ok, I finished the refactoring.

I did not at tests for the helper, as we did not have them before and I first wanted to ask, whether you think they are necessary as they indirectly tested via the service tests.

My second question is: Do you think we should shorten the Exceptions more often by introducing explicit exceptions like Unauthorized.Single.Criterion or so? In general the exceptions are very long and I do not quite like it.

Furthermore, I fixed some mistakes in the logic and tried to implement the new definition from the API documentation. But please have a second look. I did not adapt everything as it does not belong here and should be done separately.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/UserAccessRule.kt:
```diff
@@ -0,0 +1,80 @@
+val isTargetUserActive = UserAccessRule { _, target ->
+    target.status == UserStatus.USER_STATUS_ACTIVE ||
+        target.status == UserStatus.USER_STATUS_ACTIVE_UNCONFIRMED
+}
```
I am not sure how to fix this, but in this case, we don't ensure, that the `checkFor` has to be called. What do you think of: 
```kotlin
@CheckReturnValue
fun isTargetUserActive(): UserAccessRule<User> =
    UserAccessRule { _, target ->
        target.status == UserStatus.USER_STATUS_ACTIVE ||
            target.status == UserStatus.USER_STATUS_ACTIVE_UNCONFIRMED
    }
```
Same for `targetUserIsNotAdmin` and `isServerSameUser`. I know it is a bit more ugly, but would ensure the `checkFor`.

Merseleo replied:
I agree, that it is uglier, but also safer and in this case we should go with the safer version; I agree to you.

Merseleo replied:
I changed it for all rules that exist. And actually this isn't that ugly.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -143,8 +123,10 @@ class UserService(
+        isServerAdmin.forTarget<User>()
+            .orElse(isTargetUserActive)
+            .orElseThrow(UserNotFoundException(request.id))
             .checkFor(currentUser, targetUser)
```
This should be something like: 
```kotlin
isServerAdmin.forTarget<User>()
    .orElse(isTargetUserActive)
    orElseThrow(UnauthorizedException.Single(
    EntityType.USER, request.id, 
    AccessType.READ, currentUser.id.toString(), IdentifierType.ID
    ))
    .checkFor(currentUser, targetUser)
```
The exception that is thrown should be an unathorized exception imo.

Merseleo replied:
Here, I do not agree: We agreed, that an `UnauthorizedException` should be thrown in these two cases: "The calling user is neither a server admin nor a member of the same project as the requested user." But maybe I am in the same project, then I should actually see the user. 
If the user is deleted (the only case this check can fail and an exception is thrown), then in my opinion it should be treated as nonexistent anymore for all other users than the server admin (at least that was the idea).
And again, that wasn't only my idea, but as well the idea of the author of the `getUserById` function before the refactoring. If you say it should be an `UnauthorizedException` then please explain your reasons in more detail.

domienderle replied:
All right. No, it makes sense in this way too.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -98,77 +109,51 @@ class UserService(
+        isServerAdmin.forTarget<User>()
+            .orElse(isTargetUserActive)
+            .orElseThrow(UserNotFoundException(request.id))
+            .checkFor(currentUser, targetUser)
```
This should be something like: 
```kotlin
isServerAdmin.forTarget<User>()
    .orElse(isTargetUserActive)
    orElseThrow(UnauthorizedException.Single(
    EntityType.USER, request.id, 
    AccessType.READ, currentUser.id.toString(), IdentifierType.ID
    ))
    .checkFor(currentUser, targetUser)
```
The exception that is thrown should be an unathorized exception imo.

Merseleo replied:
Isn't this the same line as in the comment above?

domienderle replied:
Oh sorry, I wrote this comment twice...

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -223,22 +225,27 @@ class UserService(
+            .forProperty(User::id)
+            .andAlso(
+                targetUserIsNotAdmin
+                    .orElse(isSameUserById.forProperty(User::id))
```
Is this line necessary? If it is the same user the `isServerAdminOrSameUser` returns true and and if additionally the `targetUserIsNotAdmin` returns false, then the user must be the same user or do I miss something? I mean we could leave this check, maybe it is more obvious what the rule chaining does, but then the check would be unnecessary.

Merseleo replied:
But removing this line would indeed change the behavior. At the moment a server admin's account can only deleted by oneself. If we remove this line, then a server admin can also deletes the accounts of another server admins. I am ok with that, but it is a different behavior than it was before the refactoring.

domienderle replied:
Ah true. No then I think it is better the way it is now.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/ProjectAccessRule.kt:
```diff
@@ -0,0 +1,71 @@
+                AccessType.READ,
+                currentUser.id.toString(),
+            )
+        }
+}
+
+/**
+ * Check whether the current user is an admin of the specified project. If the user is not a project admin,
+ * the user has to be a server admin; otherwise, throws an [UnauthorizedException.Single].
+ */
+@CheckReturnValue
```
To keep consitency please add the comments for the parameters too.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -143,36 +167,32 @@ class ProjectService(
+                throw NotFoundException(EntityType.PROJECT, projectId.toString())
             }
 
+            isAllowedToReadProject(projectMemberRepo).checkFor(currentUser, projectId)
```
Please turn the ordering of this two calls around. If the user is not a server admin, he should not be able to see, that the project does not exist.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/ReviewService.kt:
```diff
@@ -44,58 +43,41 @@ interface IReviewService {
-        val projectPaper = projectPaperRepo.getProjectPaperById(review.projectPaperId).getOrThrow()
 
-        ensureCurrentUserIsProjectMember(projectPaper.projectId, currentUser)
+        isAllowedToReadReview(projectMemberRepo, projectPaperRepo).checkFor(currentUser, review)
```
Do we need to call the `checkFor` here with `review` or couldn't we just call `checkFor(currentUser, review.id)` (If we change the the `isAllowedToReadReview` to `AccessRule<UUID>`)? Then we could change the ordering of the `isAllowedToReadReview` and `getReviewById`. In this case nobody who is not allowed to know if the review exists or not would get information about this.

Merseleo replied:
But to check whether a user is allowed to read a review, we need to check whether they are in the same project (by our definition). And to check in which project the review is, I need to entire review object in the access check to get the `projectPaperId` to retrieve the `projectId`. So only parsing the `reviewId` would result in a refetch of the review inside the access check.

domienderle replied:
All right. Then we have to keep it.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/accessrules/ReviewAccessRule.kt:
```diff
@@ -0,0 +1,58 @@
+fun isAllowedToReadReview(
+    projectMemberRepo: IProjectMemberTableRepo,
+    projectPaperRepo: IProjectPaperTableRepo,
+): AccessRule<Review> {
```
In case we don't change this to `AccessRule<UUID>`, use `ReviewAccessRule` here.

Merseleo replied:
But I decided to change all things back to AccessRule<myType> and does not specify an inheriting function interface. Actually this isn't this necessary, but makes everything more complicated. Are you ok with that? Actually now everything should be consistent.

domienderle replied:
Off course. This is fine with me too. It is just about the consitency.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/InvitationService.kt:
```diff
@@ -92,28 +94,14 @@ class InvitationService(
         // Check whether a project with the given id exists
         val projectId = parseUUID(request.projectId, EntityType.PROJECT)
         val project = projectRepo.getProjectById(projectId).getOrThrow()
```
Move this line below the `isServerOrProjectAdmin` check.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/InvitationService.kt:
```diff
@@ -92,28 +94,14 @@ class InvitationService(
+            .checkFor(currentUser, project)
+
+        isServerOrProjectAdmin(projectMemberRepo, AccessType.READ)
+            .checkFor(currentUser, projectId)
```
Change the sorting of these two calls.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectPaperService.kt:
```diff
@@ -165,49 +164,37 @@ class ProjectPaperService(
+        userRepo,
+    ) { currentUser ->
         val projectId = parseUUID(request.projectId, EntityType.PROJECT)
         if (!projectRepo.doesProjectExistById(projectId)) {
             throw NotFoundException(EntityType.PROJECT, projectId.toString())
         }
 
         val relativeId = request.relativeProjectPaperId.toLong()
         val projectPaper = repo.getProjectPaperByRelativeId(projectId, relativeId).getOrThrow()
```
Move this lines below the access check. You can use `request.projectId` instead of `projectPaper.projectId` in the access check then.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectPaperService.kt:
Change the order of the access check and
```kotlin
if (!projectRepo.doesProjectExistById(projectId)) {
    throw NotFoundException(EntityType.PROJECT, projectId.toString())
}
``` 
in the `getProejctPapers` method.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectPaperService.kt.kt:
```diff
@@ -230,22 +217,27 @@ class ProjectPaperService(
+                .orElseThrow(EntityNotActiveException(EntityType.PROJECT, projectId.toString()))
+                .checkFor(currentUser, project)
+
+            isServerOrProjectAdmin(projectMemberRepo, AccessType.CREATE).checkFor(currentUser, projectId)
```
Change the order here to: 
1. `isServerAdminOrProjectAdmin`
2. `isProjectActive`
3. `doesProjectPaperExist`

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt:
```diff
@@ -71,131 +74,46 @@ class CriterionService(
-        val criterion = repo.getCriterionById(criterionId).getOrThrow()
-
-        checkCriterionPermission(criterion, currentUser, AccessType.READ)
+            .checkFor(currentUser, criterion)
```
Change the order of the access check and the `getCriterionById` call.

Merseleo replied:
This is not possible as I need the criterion object to perform the access check.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt:
```diff
@@ -71,131 +74,46 @@ class CriterionService(
-                    currentUser,
-                    AccessType.CREATE,
-                )
+                val project = projectRepo.getProjectById(parseUUID(request.projectId, EntityType.PROJECT)).getOrThrow()
```
Order:
1. `isProjectAdmin`/`isServerAdmin`
2. `isProjectActive`
3. `getProjectById`

Merseleo replied:
I cannot check whether the project is active before I retrieve the project.

domienderle commented on src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt:
```diff
@@ -206,7 +124,26 @@ class CriterionService(
             val criterion = repo.getCriterionById(criterionId).getOrThrow()
 
-            checkCriterionPermission(criterion, currentUser, AccessType.UPDATE)
+            if (criterion is Criterion.ProjectCriterion) {
+                val project = projectRepo.getProjectById(criterion.projectId).getOrThrow()
+
+                isProjectActive
+                    .orElseThrow(EntityNotActiveException(EntityType.PROJECT, project.id.toString()))
+                    .checkFor(currentUser, project)
```
Move this below the other access check.

Merseleo replied:
Same here. I need the criterion to check whether it is a user or project criterion and to perform the correct access checks. And I do not really have an idea on how to not use the criterion to determine which checks should be performed.

And i do not know on how we could improve this.

domienderle reviewed (changes requested):
In general this looks really good. I didn't find any access check that is specified in the documentation but not in your code. This is great work. There are only a couple of things that need a re-work. One thing in general:

Please have a look on the access rule classes again. Sometimes, you specify a specific access rule like `UUIDAccessRule`, but use `AccessRule<UUID>` later (same for other examples). Use the rules you explicitly specified.

Commit:
refactor: unify access rule definitions by removing redundant fun interfaces like.

For example, remove fun interface UUIDAccessRule: AccessRule<UUID>.

Commit:
refactor: convert access rule properties to functions and ensure consistent usage across services

Additionally, annotate all access rule functions with `@CheckReturnValue`.

Commit:
docs: add a missing comment to the `isServerOrProjectAdmin` project access rule

Commit:
refactor: add an overloaded ` checkFor ` function with defined target 'Unit'

This overloaded function can be used when no entity should be accessed, but in general an access should be checked.

Commit:
refactor: reorder some access rule checks

Ensure a consistent order of access checks (if possible) to not leak any details. For example, if a user is not authorized to read a project, the user should not get the information that a project doesn't exist.

Merseleo reviewed (commented):
Ok, thank you for the review and that you double checked everything! 
I hope I did not miss any of your comments and adapted everything correctly.

domienderle reviewed (approved):
Really, really good. I really don't find anything else to improve. This is good work and makes our code a lot cleaner. Thanks for the refactoring. You can merge now.
